kernel/byterun/coq_interp.c

```diff
@@ -1771,6 +1773,63 @@ value coq_interprete
         Next;
       }
 
+
+      Instruct(ISINT_CAML_CALL2) {
+              print_instr("ISINT_CAML_CALL2");
+              if (Is_uint63(accu)) {
+                pc++;
+                print_int(*pc);
+                Setup_for_caml_call;
+                accu = caml_callback2(Field(coq_global_data, *pc),accu, sp[1]);
+                Restore_after_caml_call;
+                sp += 1;
+                pc++;
+              } else pc += *pc;
+              Next;
+      }
+
+      Instruct(ISARRAY_CAML_CALL1) {
+        print_instr("ISARRAY_CAML_CALL1");
+              if (Is_coq_array(accu)) {
```

Comment:
The naming `ISARRAY_CAML_CALL1` and `Is_coq_array` is really confusing, because what is actually tested is a length-1 block. It does not have anything to do with arrays in the first place.

More generally, this whole callback mechanism does not seem to care whether things are integers or blocks in the first place (and even less about the size of these blocks). The only thing that matters is whether some arguments are possibly accumulators.

So, it would seem better to systematically use `!Is_accu(...)` and to name the opcodes something like `ISNONACCU2_CALL3` to mean that this is a 3-argument callback that is valid as soon as the first two arguments are not accumulators.

Reply:
Unfortunately, `!Is_accu` does not work, since persistent arrays are references, they seem to carry tag `0` which is also the tag of accumulators (that's why we use the size to distinguish both). Of course, I'm happy to refine the naming if there are other suggestions.

Reply:
I thought accumulators were using a different tag that could not conflict. That is really unfortunate.